cpp/src/stream_compaction/distinct.cu

```diff
@@ -57,55 +230,21 @@ std::unique_ptr<table> distinct(table_view const& input,
     return empty_like(input);
   }
 
-  auto keys_view = input.select(keys);
-  auto preprocessed_keys =
-    cudf::experimental::row::hash::preprocessed_table::create(keys_view, stream);
-  auto const has_null = nullate::DYNAMIC{cudf::has_nested_nulls(keys_view)};
-  auto const num_rows{keys_view.num_rows()};
-
-  hash_map_type key_map{compute_hash_table_size(num_rows),
-                        cuco::sentinel::empty_key{COMPACTION_EMPTY_KEY_SENTINEL},
-                        cuco::sentinel::empty_value{COMPACTION_EMPTY_VALUE_SENTINEL},
-                        detail::hash_table_allocator_type{default_allocator<char>{}, stream},
-                        stream.value()};
-
-  auto row_hash = cudf::experimental::row::hash::row_hasher(preprocessed_keys);
-  experimental::compaction_hash hash_key(row_hash.device_hasher(has_null));
-
-  cudf::experimental::row::equality::self_comparator row_equal(preprocessed_keys);
-  auto key_equal = row_equal.equal_to(has_null, nulls_equal);
-
-  auto iter = cudf::detail::make_counting_transform_iterator(
-    0, [] __device__(size_type i) { return cuco::make_pair(i, i); });
-  // insert distinct indices into the map.
-  key_map.insert(iter, iter + num_rows, hash_key, key_equal, stream.value());
-
-  auto const output_size{key_map.get_size()};
-  auto distinct_indices = cudf::make_numeric_column(
-    data_type{type_id::INT32}, output_size, mask_state::UNALLOCATED, stream, mr);
-  // write distinct indices to a numeric column
-  key_map.retrieve_all(distinct_indices->mutable_view().begin<cudf::size_type>(),
-                       thrust::make_discard_iterator(),
-                       stream.value());
-
-  // run gather operation to establish new order
-  return detail::gather(input,
-                        distinct_indices->view(),
-                        out_of_bounds_policy::DONT_CHECK,
-                        detail::negative_index_policy::NOT_ALLOWED,
-                        stream,
-                        mr);
+  auto const gather_map = get_distinct_indices(input, keys, keep, nulls_equal, stream);
+  return detail::gather(
+    input, gather_map.begin(), gather_map.end(), out_of_bounds_policy::DONT_CHECK, stream, mr);
```

Comment:
Can you call this gather instead?
https://github.com/rapidsai/cudf/blob/ac2d9a6fc84e9b74e3fc95e97dfddbeea3d54961/cpp/include/cudf/detail/gather.hpp#L78-L84
It will accept the `device_uvector` and automatically convert it to a `device_span`. This way you can include `gather.hpp` instead of `gather.cuh` and it may improve compile time.